test/unit-test/gtclang/Parsing/TestParsing.cpp

```diff
@@ -0,0 +1,175 @@
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "dawn/CodeGen/CXXUtil.h"
+#include "dawn/SIR/SIR.h"
+#include "gtclang/Support/ParsingHeader.h"
+#include "gtclang/Unittest/GTClang.h"
+#include "gtclang/Unittest/UnittestEnvironment.h"
+#include <fstream>
+#include <gtest/gtest.h>
+#include <map>
+
+using namespace dawn;
+using namespace gtclang;
+
+namespace {
+
+struct NestableFunctions : public codegen::MemberFunction {
+  NestableFunctions(const Twine& type, const Twine& name, std::stringstream& s, int il = 0)
+      : MemberFunction(type, name, s, il) {}
+
+  NestableFunctions addFunction(const Twine& returntype, const Twine& name) {
+    NestableFunctions nf(returntype, name, ss(), IndentLevel + 1);
+    return nf;
+  }
+
+  NestableFunctions addVerticalRegion(const Twine& min, const Twine& max) {
+    NestableFunctions nf(Twine::createNull(), "vertical_region", ss(), IndentLevel + 1);
+    nf.addArg(min).addArg(max);
+    nf.startBody();
+    return nf;
+  }
+
+  NestableFunctions addIfStatement(const Twine& ifCondition) {
+    NestableFunctions ifStatement(Twine::createNull(), "if", ss(), IndentLevel + 1);
+    ifStatement.addArg(ifCondition);
+    ifStatement.startBody();
+    return ifStatement;
+  }
+};
+
+struct VerticalRegion : public NestableFunctions {
+  VerticalRegion(const Twine& name, std::stringstream& s, int il = 0)
+      : NestableFunctions(Twine::createNull(), name, s, il) {}
+};
+
+struct DawnObject : public codegen::Structure {
+  using Structure::Structure;
+  DawnObject(const Twine& type, const Twine& name, std::stringstream& s)
+      : Structure(type.str().c_str(), name, s) {}
+
+  Statement addStorage(const Twine& memberName) {
+    Statement member(ss(), IndentLevel + 1);
+    member << "storage"
+           << " " << memberName;
+    return member;
+  }
+
+  NestableFunctions addDoMethod(const Twine& type) {
+    NestableFunctions nf(type, "Do", ss(), IndentLevel + 1);
+    return nf;
+  }
+
+  Statement addOffset(const Twine& offsetName) {
+    Statement member(ss(), IndentLevel + 1);
+    member << "offset"
+           << " " << offsetName;
+    return member;
+  }
+};
+
+struct StencilFunction : public DawnObject {
+  StencilFunction(const Twine& name, std::stringstream& s)
+      : DawnObject("stencil_function", name, s) {}
+
+  virtual NestableFunctions addDoMethod(const Twine& type) {
+    NestableFunctions nf(type, "Do", ss(), IndentLevel + 1);
+    nf.startBody();
+    return nf;
+  }
+};
+
+struct Stencil : public DawnObject {
+  Stencil(const Twine& name, std::stringstream& s) : DawnObject("stencil", name, s) {}
+};
+
+struct Globals : public DawnObject {
+  Globals(std::stringstream& s) : DawnObject("globals", Twine::createNull(), s) {}
+  virtual NestableFunctions addDoMethod(const Twine& type) = delete;
+  virtual Statement addOffset(const Twine& offsetName) = delete;
+};
+
+class FileWriter {
```

Comment:
Should be in a separate file. The convention is usually to name the file the same as the class it contains (at least for the header).